Patch `utcnow` in retry delay test

edwardwang888 opened the pull request:
This test uses the current timestamp and sleep to test `retry_delay`. This is nondeterministic and fails on my local machine as it takes longer than 3 seconds to execute the first two runs. The `retry_delay` code was originally written using `datetime.datetime.now`, which is a built-in method and probably could not be easily patched without affecting other parts of the code that used `datetime.datetime`. However, now that `timezone.utcnow` is used, patching is possible. The logic of `timezone.utc` does need to be duplicated, though it is only two lines.
https://github.com/apache/airflow/blob/87e98652d58198e4ef7146bdc7e1e8d87a7d2bae/airflow/utils/timezone.py#L53-L65

---
Original commit that added these tests: https://github.com/apache/airflow/commit/8051aa988d315ab5c54bf5dd8f6e931356c07c2e#diff-969c0712020e8472d8d78ae32190692cc1e8a3e16dc17f6977c8f87a1328adb7R142

At that time, `retry_delay` logic was done using `datetime.datetime.now`: https://github.com/apache/airflow/blob/8051aa988d315ab5c54bf5dd8f6e931356c07c2e/airflow/models.py#L928-L934
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

edwardwang888 committed:
Remove sleep from retry delay test

mik-laj commented on tests/models/test_taskinstance.py:
```diff
@@ -17,15 +17,16 @@
 from unittest import mock
 from unittest.mock import call, mock_open, patch
 
+import numpy as np
```
This library is not an Airflow dependency.  If we want to use it for core testing, it should be installed with `devel` extra.
https://github.com/apache/airflow/blob/main/setup.py#L495
Or maybe you have an idea how to write this test without using this library?

edwardwang888 replied:
Yeah, it is not strictly necessary. I was just trying to produce the closest boundary case 🙂 

In general, is there a preference of whether to add (not strictly necessary) libraries to the `devel` extra?
https://github.com/apache/airflow/blob/3f65a48bcedfab6a50e24ecac710170aab6f51bc/tests/models/test_taskinstance.py#L526-L532

edwardwang888 replied:
Actually, I just saw #12500. It doesn't seem necessary to require `numpy` just for this test, so I will change it. But do you know why the test still passed in the CI environment?

EDIT: It looks like `pandas` is still a library in `devel_ci`. I think that's why.

mik-laj replied:
We have a lot of transitive dependencies. In this case, we have installed `numpy` on CI, because. it is required by `apache-beam`, `pandas` and `pyarrow`.
```
$ pipdeptree -r -p numpy
numpy==1.19.5
  - apache-beam==2.32.0 [requires: numpy>=1.14.3,<1.21.0]
  - pandas==1.1.5 [requires: numpy>=1.15.4]
    - pandas-gbq==0.14.1 [requires: pandas>=0.20.1]
    - scrapbook==0.5.0 [requires: pandas]
  - pyarrow==4.0.1 [requires: numpy>=1.16.6]
    - apache-beam==2.32.0 [requires: pyarrow>=0.15.1,<5.0.0]
    - scrapbook==0.5.0 [requires: pyarrow]
```
Airflow does not depend on these dependencies, so if we want to use this library, we should add it.
```
$ curl -s https://raw.githubusercontent.com/apache/airflow/main/setup.{py,cfg} | grep -C 10 numpy | wc -l
0
```
> In general, is there a preference of whether to add (not strictly necessary) libraries to the devel extra?

In this case, we should add this library to `devel`, because itis needed for core.

extra | command | description
-- | -- | --
devel | pip install 'apache-airflow[devel]' | Minimum dev tools requirements (without providers)
devel_hadoop | pip install 'apache-airflow[devel_hadoop]' | Same as devel + dependencies for developing the Hadoop stack
devel_all | pip install 'apache-airflow[devel_all]' | Everything needed for development (devel_hadoop + providers)
devel_ci | pip install 'apache-airflow[devel_ci]' | All dependencies required for CI build.

See: http://airflow.apache.org/docs/apache-airflow/stable/extra-packages-ref.html#bundle-extras

mik-laj replied:
I also support the idea of getting rid of this dependency, but if you don't find another solution then it's okay.

edwardwang888 replied:
Thanks for the explanation! Yes, I found a solution where the difference is practically negligible, so I will just change it and remove `numpy` from the test.

mik-laj commented on tests/models/test_taskinstance.py:
```diff
@@ -489,10 +490,14 @@ def test_task_retry_wipes_next_fields(self, session, dag_maker):
         assert ti.state == State.UP_FOR_RETRY
 
-    def test_retry_delay(self, dag_maker):
+    @patch('airflow.utils.timezone.utcnow')
```
What do you think about [freezegun](https://github.com/spulec/freezegun)?

edwardwang888 replied:
Cool, looks like exactly what I need!

edwardwang888 committed:
Change to use freezegun

edwardwang888 committed:
Approximate smallest timedelta without numpy

edwardwang888 committed:
Use pre-defined smallest timedelta

It turns out `datetime.datetime.resolution` is already defined as the smallest possible timedelta (one microsecond). Attempting to create a timedelta with microseconds in the interval (0.5, 1) simply results in microseconds being rounded up to 1.
